pkg/kube-metrics/metrics.go

```diff
@@ -0,0 +1,94 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubemetrics
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	kcollector "k8s.io/kube-state-metrics/pkg/collector"
+	ksmetric "k8s.io/kube-state-metrics/pkg/metric"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+var log = logf.Log.WithName("kubemetrics")
+
+// ServeCRMetrics generates CustomResource specific metrics based on the inputted custom resource GVK.
+// A list of namespaces, ns, can be passed to ServeCRMetrics to scope the generated metrics. Passing nil or
+// an empty list of namespaces will result in an error.
+// The function also starts serving the generated collections of the metrics on given host and port.
+func ServeCRMetrics(cfg *rest.Config,
+	ns []string,
+	operatorGVKs []schema.GroupVersionKind,
+	host string, port int32) error {
+	// We have to have at least one namespace.
+	if len(ns) < 1 {
+		return errors.New("namespaces were empty; pass at least one namespace to generate custom resource metrics")
+	}
+	// Create new unstructured client.
+	var collectors [][]kcollector.Collector
+	log.V(1).Info("Starting collecting operator types")
+	// Loop through all the possible operator/custom resource specific types.
+	for _, gvk := range operatorGVKs {
+		apiVersion := gvk.GroupVersion().String()
+		kind := gvk.Kind
+		// Generate metric based on the kind.
+		metricFamilies := generateMetricFamilies(gvk.Kind)
+		log.V(1).Info("Generating metric families", "apiVersion", apiVersion, "kind", kind)
+		dclient, err := newClientForGVK(cfg, apiVersion, kind)
+		if err != nil {
+			return err
+		}
+		// Generate collector based on the group/version, kind and the metric families.
+		c := NewCollectors(dclient, ns, apiVersion, kind, metricFamilies)
+		collectors = append(collectors, c)
+	}
+	// Start serving metrics.
+	log.V(1).Info("Starting serving custom resource metrics")
+	go ServeMetrics(collectors, host, port)
+
+	return nil
+}
+
+func generateMetricFamilies(kind string) []ksmetric.FamilyGenerator {
+	helpText := fmt.Sprintf("Information about the %s custom resource .", kind)
+	kindName := fmt.Sprintf("%s", strings.ToLower(kind))
+	metricName := fmt.Sprintf("%s_info", strings.ToLower(kind))
+
+	return []ksmetric.FamilyGenerator{
+		ksmetric.FamilyGenerator{
+			Name: metricName,
+			Type: ksmetric.Gauge,
+			Help: helpText,
+			GenerateFunc: func(obj interface{}) *ksmetric.Family {
+				crd := obj.(*unstructured.Unstructured)
+				return &ksmetric.Family{
+					Metrics: []*ksmetric.Metric{
+						{
+							Value:       1,
+							LabelKeys:   []string{"namespace", kindName},
```

Comment:
@LiliC Is the second label supposed to be the kind name e.g `memcached=app1` or is it supposed to be the string `name` e.g `name=app1`.
Going by your previous should this show as:
```console
# HELP jonthebeach_info Information about the JOnTheBeach operator custom resource replica.
# TYPE jonthebeach_info gauge
jonthebeach_info{namespace="default",name="example-jonthebeach"} 1
jonthebeach_info{namespace="default",name="example-jonthebeach2"} 1
```

Reply:
it's supposed to be `memcached=app1 ` as seen above:
```
# HELP jonthebeach_info Information about the JOnTheBeach operator custom resource replica.
# TYPE jonthebeach_info gauge
jonthebeach_info{namespace="default",jonthebeach="example-jonthebeach"} 1
jonthebeach_info{namespace="default",jonthebeach="example-jonthebeach2"} 1
```